packages/airnode-protocol/contracts/WIP/Airnode.sol

```diff
@@ -0,0 +1,273 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.9;
+
+import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+import "@openzeppelin/contracts/utils/Multicall.sol";
+import "./WithdrawalUtils.sol";
+import "./interfaces/IAirnode.sol";
+
+contract Airnode is Multicall, WithdrawalUtils, IAirnode {
+    using ECDSA for bytes32;
+
+    struct Template {
+        address airnode;
+        bytes32 endpointId;
+        bytes parameters;
+    }
+
+    mapping(bytes32 => Template) public templates;
+
+    mapping(address => mapping(address => bool))
+        public sponsorToRequesterToSponsorshipStatus;
+
+    mapping(address => uint256) public requesterToRequestCountPlusOne;
+
+    mapping(bytes32 => bytes32) private requestIdToFulfillmentParameters;
+
+    function createTemplate(
+        address airnode,
+        bytes32 endpointId,
+        bytes calldata parameters
+    ) external returns (bytes32 templateId) {
+        templateId = keccak256(
+            abi.encodePacked(airnode, endpointId, parameters)
+        );
+        templates[templateId] = Template({
+            airnode: airnode,
+            endpointId: endpointId,
+            parameters: parameters
+        });
+        emit CreatedTemplate(templateId, airnode, endpointId, parameters);
+    }
+
+    function setSponsorshipStatus(address requester, bool sponsorshipStatus)
+        external
+    {
+        if (requesterToRequestCountPlusOne[requester] == 0) {
+            requesterToRequestCountPlusOne[requester] = 1;
+        }
+        sponsorToRequesterToSponsorshipStatus[msg.sender][
+            requester
+        ] = sponsorshipStatus;
+        emit SetSponsorshipStatus(msg.sender, requester, sponsorshipStatus);
+    }
+
+    function makeRequest(
+        bytes32 templateId,
+        address reporter,
+        address sponsor,
+        address sponsorWallet,
+        address fulfillAddress,
+        bytes4 fulfillFunctionId,
+        bytes calldata parameters
+    ) external returns (bytes32 requestId) {
+        require(fulfillAddress != address(this), "Fulfill address AirnodeRrp");
+        require(
+            sponsorToRequesterToSponsorshipStatus[sponsor][msg.sender],
+            "Requester not sponsored"
+        );
+        uint256 requesterRequestCount = requesterToRequestCountPlusOne[
+            msg.sender
+        ];
+        requestId = keccak256(
+            abi.encodePacked(
+                block.chainid,
+                address(this),
+                msg.sender,
+                requesterRequestCount,
+                templateId,
+                reporter,
+                sponsor,
+                sponsorWallet,
+                fulfillAddress,
+                fulfillFunctionId,
+                parameters
+            )
+        );
+        requestIdToFulfillmentParameters[requestId] = keccak256(
+            abi.encodePacked(
+                templates[templateId].airnode,
+                reporter,
```

Comment:
We need to add this to the request ID calculation because it's used in `fulfill()`